### ISSUE TITLE
Helix JobSender SDK doesn't provide a way to set IsExternal=true

### ISSUE THREAD

**Issue by akoeplinger** [dotnet/arcade#1392]
E.g. using this Hello World sample:

```csharp
var api = ApiFactory.GetAuthenticated("<API Token>");

var jobDefinition = api.Job.Define()
        .WithSource("pr/test/helloworld")
        .WithType("test/helloworld")
        .WithBuild(DateTime.UtcNow.Ticks.ToString())
        .WithTargetQueue("Ubuntu.1804.Amd64.Open")
        .DefineWorkItem("Hello World")
        .WithCommand("echo 'Hello!'")
        .WithEmptyPayload()
        .AttachToJob();

var job = await jobDefinition.SendAsync();
```

The resulting build isn't accessible to users outside of the Microsoft GitHub org on Mission Control, @MattGal says this is because the Helix job has IsExternal=false but there seems to be no way to set this to true with the SDK.

**Comment by akoeplinger** [comment]
Btw. this is a pretty important issue for us since right now external contributors can't view their PR test results on MC.

**Comment by akoeplinger** [comment]
Thanks.

Re: AzDO, we're not there yet and won't be for some time :)

**Comment by akoeplinger** [comment]
Ok so according to MattGal the only requirements are to 1) provide a 'Source" value that starts with pr/ or automated/ and 2) not be authenticated when submitting.

The issue in my case is 2) as I'm using the authenticated API instead of anonymous. The problem is that I can't switch over to anonymous since then the SDK will throw a "not authorized" exception when trying to upload the payload to the Helix-provided Azure blob storage (which of course requires auth).

It seems that the SDK should allow providing two different sets of credentials to support this.

@Chrisboh I assume you use the older method of manually uploading the payload to Azure and just providing the payload URI?

**Comment by Chrisboh** [comment]
@markwilkie this is a bug with the current SDK which wasn't caught yet as most people are logged in by default.  This is something we should add to Jon's work for Helix. 

@akoeplinger in the meantime can you have the person you are working with join the Microsoft org so he can be unblocked while we work on this?

**Comment by akoeplinger** [comment]
@Chrisboh Yup, that works for Microsoft contributors but not for external community.

**Comment by akoeplinger** [comment]
I found a workaround that seems to work for my use case:

```csharp
var authenticatedApi = ApiFactory.GetAuthenticated("<API TOKEN>");
var storage = new Storage((HelixApi)authenticatedApi);

var anonymousApi = ApiFactory.GetAnonymous();
typeof(HelixApi).GetProperty("Storage").SetValue(anonymousApi, storage, null);

var jobDefinition = anonymousApi.Job.Define()
...
```

**Comment by akoeplinger** [comment]
@jonfortescue I don't see a fix for this. https://github.com/dotnet/arcade/pull/1451 isn't it since that only adds the workaround to the MSBuild SDK, not an actual proper API for the JobSender SDK.

**Comment by jonfortescue** [comment]
@akoeplinger just to clarify -- are you wanting to have a first class method (a la `AuthenticatedApi.DefineExternal()`)? Or do you want the change to happen at the Helix API level -- e.g. setting a property and having Helix know how to make the job external?

**Comment by alexperovich** [comment]
A settable property in the helix api would probably be easiest. Though we could probably make a parameter to `SendAsync` that will do the same thing.

**Comment by akoeplinger** [comment]
I don't really have any preference - my point was just that using reflection to mess with a private property is not a real solution :)

I think a property on the Helix API is fine like @alexperovich mentioned. Not sure I'd put it on SendAsync as other parts of the SDK might need authenticated access too in the future and then we need to add a parameter everywhere.

**Comment by jonfortescue** [comment]
With the change to the storage APIs, this is obviated.

**Comment by akoeplinger** [comment]
@jonfortescue can you please point me to the PR where this was fixed? does `ApiFactory.GetAuthenticated()` now work for external jobs as well?

**Comment by alexperovich** [comment]
Any jobs sent with ApiFactory.GetAnonymous() are external. It is not possible to send work to the external queues if you are authenticated.

________________________________

**Comment by akoeplinger** [comment]
@alexperovic sure, but the problem in this issue was that you couldn't upload to the helix-provided storage container using the anonymous API. I assume that's fixed now?

**Comment by alexperovich** [comment]
Yes, the anonymous api allows limited storage access that works for 99% of helix jobs. IIRC the storage api only allows getting a new container and it limits the tokens to 1 week validity. @MattGal is this correct?

**Comment by akoeplinger** [comment]
Ok thanks, I was able to send jobs with the new package but hit another issue: https://github.com/dotnet/arcade/issues/1807
